worldengine/draw.py

```diff
@@ -201,34 +268,424 @@ def draw_riversmap(world, target):
     sea_color = (255, 255, 255, 255)
     land_color = (0, 0, 0, 255)
 
-    for y in range(world.height):
+    for y in range(world.height):#TODO: numpy
         for x in range(world.width):
-            target.set_pixel(x, y, sea_color if world.is_ocean((x, y)) else land_color)
+            if world.ocean[y, x]:
+                target.set_pixel(x, y, sea_color)
+            else:
+                target.set_pixel(x, y, land_color)
 
     draw_rivers_on_image(world, target, factor=1)
 
 
 def draw_grayscale_heightmap(world, target):
-    e = world.elevation['data']
+    min_elev_sea = None
+    max_elev_sea = None
+    min_elev_land = None
+    max_elev_land = None
+    for y in range(world.height):
+        for x in range(world.width):
+            e = world.elevation['data'][y, x]
+            if world.is_land((x, y)):
+                if min_elev_land is None or e < min_elev_land:
+                    min_elev_land = e
+                if max_elev_land is None or e > max_elev_land:
+                    max_elev_land = e
+            else:
+                if min_elev_sea is None or e < min_elev_sea:
+                    min_elev_sea = e
+                if max_elev_sea is None or e > max_elev_sea:
+                    max_elev_sea = e
 
-    mask = numpy.ma.array(e, mask=world.ocean)  # only land
-    min_elev_land = mask.min()
-    max_elev_land = mask.max()
     elev_delta_land = max_elev_land - min_elev_land
-
-    mask = numpy.ma.array(e, mask=numpy.logical_not(world.ocean))  # only ocean
-    min_elev_sea = mask.min()
-    max_elev_sea = mask.max()
     elev_delta_sea = max_elev_sea - min_elev_sea
 
-    c = numpy.empty(e.shape, dtype=numpy.float)
-    c[numpy.invert(world.ocean)] = (e[numpy.invert(world.ocean)] - min_elev_land) * 127 / elev_delta_land + 128
-    c[world.ocean] = (e[world.ocean] - min_elev_sea) * 127 / elev_delta_sea
-    c = numpy.rint(c).astype(dtype=numpy.int32)  # proper rounding
-
     for y in range(world.height):
         for x in range(world.width):
-            target.set_pixel(x, y, (c[y, x], c[y, x], c[y, x], 255))
+            e = world.elevation['data'][y, x]
+            if world.is_land((x, y)):
+                c = int(((e - min_elev_land) * 127) / elev_delta_land)+128
+            else:
+                c = int(((e - min_elev_sea) * 127) / elev_delta_sea)
+            target.set_pixel(x, y, (c, c, c, 255))
+
+
+def draw_satellite(world, target):
+    # Not sure what this first part is doing - copied from draw_greyscale_heightmap 
+    # I believe it is normalizing the heightmap values between 1 and 255?
+    # This function will use those values later, so this loop is kept in
+    min_elev_sea = None
+    max_elev_sea = None
+    min_elev_land = None
+    max_elev_land = None
```

Comment:
You should copy this code from the most recent WorldEngine-code instead since there is a much faster way of doing this now.